File: java/rsa.java
public class rsa {

    // -----------------------------
    // Step 1: RSA encryption/decryption functions
    // -----------------------------
    public static int encryptRSA(int m, int e, int n) {
        return (int) (Math.pow(m, e) % n);
    }

    public static int decryptRSA(int c, int d, int n) {
        return (int) (Math.pow(c, d) % n);
    }

    // -----------------------------
    // Step 2: Simple XOR for symmetric encryption
    // -----------------------------
    public static int[] xorEncrypt(String message, int key) {
        int[] encrypted = new int[message.length()];
        for (int i = 0; i < message.length(); i++) {
            encrypted[i] = message.charAt(i) ^ key;
        }
        return encrypted;
    }

    public static String xorDecrypt(int[] encrypted, int key) {
        char[] decrypted = new char[encrypted.length];
        for (int i = 0; i < encrypted.length; i++) {
            decrypted[i] = (char) (encrypted[i] ^ key);
        }
        return new String(decrypted);
    }

    public static void main(String[] args) {
        // -----------------------------
        // Step 3: Define keys and message
        // -----------------------------
        int receiverPublicE = 7;
        int receiverPrivateD = 3;
        int receiverN = 33;

        int symmetricKey = 4; // simple number as symmetric key
        String message = "HELLO";

        System.out.println("Original message: " + message);

        // -----------------------------
        // Step 4: Encrypt message with symmetric key
        // -----------------------------
        int[] encryptedMessage = xorEncrypt(message, symmetricKey);
        System.out.print("Encrypted message (XOR): ");
        for (int c : encryptedMessage)
            System.out.print(c + " ");
        System.out.println();

        // -----------------------------
        // Step 5: Encrypt symmetric key with receiver's public key (RSA)
        // -----------------------------
        int encryptedSymKey = encryptRSA(symmetricKey, receiverPublicE, receiverN);
        System.out.println("Encrypted symmetric key (RSA): " + encryptedSymKey);

        // -----------------------------
        // Step 6: Receiver decrypts symmetric key
        // -----------------------------
        int decryptedSymKey = decryptRSA(encryptedSymKey, receiverPrivateD, receiverN);
        System.out.println("Decrypted symmetric key: " + decryptedSymKey);

        // -----------------------------
        // Step 7: Receiver decrypts message
        // -----------------------------
        String decryptedMessage = xorDecrypt(encryptedMessage, decryptedSymKey);
        System.out.println("Decrypted message: " + decryptedMessage);
    }
}

File: java/rsa.py
from cryptography.hazmat.primitives.asymmetric import rsa, padding
from cryptography.hazmat.primitives import serialization, hashes

# -----------------------------
# Step 1: Generate receiver RSA keys
# -----------------------------
receiver_key = rsa.generate_private_key(
    public_exponent=65537,
    key_size=2048
)
receiver_public_key = receiver_key.public_key()
key=rsa.generate_private_key()
# -----------------------------
# Step 2: Simple XOR for symmetric encryption
# -----------------------------
def xor_encrypt(message, key):
    return [ord(ch) ^ key for ch in message]

def xor_decrypt(encrypted, key):
    return "".join(chr(c ^ key) for c in encrypted)

# -----------------------------
# Step 3: Hybrid Encryption Demo
# -----------------------------
message = "HELLO"
symmetric_key = 42  # simple number for XOR

print("Original message:", message)

# Encrypt message using XOR
encrypted_message = xor_encrypt(message, symmetric_key)
print("Encrypted message (XOR):", encrypted_message)

# Encrypt symmetric key using receiver's public RSA key
sym_key_bytes = symmetric_key.to_bytes(2, byteorder='big')
encrypted_sym_key = receiver_public_key.encrypt(
    sym_key_bytes,
    padding.OAEP(
        mgf=padding.MGF1(algorithm=hashes.SHA256()),
        algorithm=hashes.SHA256(),
        label=None
    )
)
print("Encrypted symmetric key (RSA):", encrypted_sym_key.hex())

# Decrypt symmetric key using receiver's private RSA key
decrypted_sym_key_bytes = receiver_key.decrypt(
    encrypted_sym_key,
    padding.OAEP(
        mgf=padding.MGF1(algorithm=hashes.SHA256()),
        algorithm=hashes.SHA256(),
        label=None
    )
)
decrypted_sym_key = int.from_bytes(decrypted_sym_key_bytes, byteorder='big')
print("Decrypted symmetric key:", decrypted_sym_key)

# Decrypt message using XOR with decrypted symmetric key
decrypted_message = xor_decrypt(encrypted_message, decrypted_sym_key)
print("Decrypted message:", decrypted_message)
